lib/chef/resource/macos_pkg.rb

```diff
@@ -0,0 +1,111 @@
+#
+# Author:: Joshua Timberman (<jtimberman@chef.io>)
+# Author:: William Theaker (<william.theaker+chef@gusto.com>)
+# Copyright:: Copyright (c) Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../resource"
+
+class Chef
+  class Resource
+    class MacosPkg < Chef::Resource
+      provides(:macos_pkg) { true }
+
+      description "Use the **macos_pkg** resource to install a macOS `.pkg` file, optionally downloading it from a remote source. A `package_id` property must be provided for idempotency. Either a `file` or `source` property is required."
+      introduced "18.1"
+      examples <<~DOC
+        **Install osquery**:
+
+        ```ruby
+        macos_pkg 'osquery' do
+          checksum   '1fea8ac9b603851d2e76c5fc73138a468a3075a3002c8cb1fd7fff53b889c4dd'
+          package_id 'io.osquery.agent'
+          source     'https://pkg.osquery.io/darwin/osquery-5.8.2.pkg'
+          action     :install
+        end
+        ```
+      DOC
+
+      allowed_actions :install
+      default_action  :install
+
+      property :checksum, String,
+        description: "The sha256 checksum of the `.pkg` file to download."
+
+      property :file, String,
+        description: "The absolute path to the `.pkg` file on the local system."
+
+      property :headers, Hash,
+        description: "Allows custom HTTP headers (like cookies) to be set on the `remote_file` resource.",
+        desired_state: false
+
+      property :package_id, String,
+        description: "The package ID registered with `pkgutil` when a `pkg` or `mpkg` is installed.",
+        required: true
+
+      property :source, String,
+        description: "The remote URL used to download the `.pkg` file."
+
+      property :target, String,
+        description: "The device to install the package on.",
+        default: "/"
+
+      load_current_value do |new_resource|
+        if shell_out("pkgutil --pkg-info '#{new_resource.package_id}'").exitstatus == 0
+          Chef::Log.debug "#{new_resource.package_id} is already installed. To upgrade, try \"sudo pkgutil --forget '#{new_resource.package_id}'\""
+        else
```

Comment:
It seems odd to have an :install action that then tells people to run manual commands. We probably want an `:upgrade` action to do something sensible here.

Reply:
ditto for this line https://github.com/chef/chef/blob/main/lib/chef/resource/dmg_package.rb#L119

Reply:
Would adding an `:upgrade` action and implicitly forgetting the install receipt by running `pkgutil --forget` in the resource be an unexpected behavior for users?

Reply:
If it will not leave cruft from the previous version, that's probably fine. Otherwise, I would uninstall as indicated above before doing the install.